src/DependencyInjection/Compiler/TemplateRegistryCompilerPass.php

```diff
@@ -0,0 +1,195 @@
+
+                continue;
+            }
+
+            $templateRegistryDefinition->addMethodCall('setTemplate', [$type, $template]);
+        }
+    }
+
+    /**
+     * @param array<string, string> $templates
+     */
+    private function setTemplatesByArgument(ContainerBuilder $container, Definition $templateRegistryDefinition, array $templates): void
+    {
+        $arguments = $templateRegistryDefinition->getArguments();
+        $serviceTemplates = $arguments[0] ?? [];
+
+        if (\is_string($serviceTemplates)) {
+            $serviceTemplates = $container->getParameter($serviceTemplates);
+        }
+
+        $templateRegistryDefinition->setArgument(0, array_merge($serviceTemplates, $templates));
+    }
+
+    private function getTemplateRegistryDefinition(ContainerBuilder $container, string $id): Definition
+    {
+        $definition = $container->getDefinition($id);
+
+        if (is_a($definition->getClass(), TemplateRegistryInterface::class, true)) {
+            return $definition;
+        }
+
+        if ($templateRegistryDefinition = $this->getTemplateRegistryFromAwareDefinition($container, $definition, $id)) {
+            return $templateRegistryDefinition;
+        }
+
+        return $this->generateTemplateRegistry($container, $definition, $id);
+    }
+
+    /**
+     * @phpstan-return class-string
+     */
+    private function getTemplateRegistryInterface(Definition $awareDefinition, string $id): string
+    {
+        $templateRegistryClass = null;
+        if (is_a($awareDefinition->getClass(), MutableTemplateRegistryAwareInterface::class, true)) {
+            $templateRegistryClass = MutableTemplateRegistryInterface::class;
+        } elseif (is_a($awareDefinition->getClass(), TemplateRegistryAwareInterface::class, true)) {
+            $templateRegistryClass = TemplateRegistryInterface::class;
+        }
+
+        if (null === $templateRegistryClass) {
+            throw new \Exception(sprintf(
+                'Tagged "%s" service must be instance of %s or %s.',
+                $id,
+                TemplateRegistryAwareInterface::class,
+                MutableTemplateRegistryAwareInterface::class
+            ));
+        }
+
+        return $templateRegistryClass;
+    }
+
+    /**
+     * @phpstan-return class-string
+     */
+    private function getTemplateRegistryFromAwareDefinition(ContainerBuilder $container, Definition $awareDefinition, string $id): ?Definition
+    {
+        $templateRegistryInterface = $this->getTemplateRegistryInterface($awareDefinition, $id);
+
+        foreach ($awareDefinition->getMethodCalls() as [$method, $args]) {
+            if ('setTemplateRegistry' === $method) {
+                $reference = $args[0];
+                $templateRegistryDefinition = $container->getDefinition($reference->__toString());
+
+                if (!is_a($templateRegistryDefinition->getClass(), $templateRegistryInterface, true)) {
+                    throw new \Exception(sprintf(
```

Comment:
IMO, this exception could use a less generic type (something like `\InvalidArgumentException`, `\BadMethodCallException` or similar.
I don't know if we have domain specific exceptions available under the `Symfony\Component\DependencyInjection\` NS for cases like this.